--- Lib/logging/__init__.py	2014-05-19 07:19:38.000000000 +0200
+++ Lib/logging/__init__.py	2014-06-10 23:30:06.551728955 +0200
@@ -53,10 +53,14 @@ __date__    = "07 February 2010"
 
 #
 # _srcfile is used when walking the stack to check when we've got the first
 # caller stack frame.
 #
+
+# Work-around for pyrun:
+__file__ = '<pyrun>/logging/__init__.pyc'
+
 if hasattr(sys, 'frozen'): #support for py2exe
     _srcfile = "logging%s__init__%s" % (os.sep, __file__[-4:])
 else:
     _srcfile = __file__
 _srcfile = os.path.normcase(_srcfile)
